Add an 'area' term for Template Parts.

Addison-Stavlo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->
This PR explores adding an 'area' (im open to better naming suggestions) term to aid in being able to classify and request all template parts corresponding to a specific type ('header', 'footer', etc.) - something like this will be necessary to enable items discussed in #27337

Current state of this PR:
* Registers the new tax term as `wp_template_part_area`.
* Allows themes to set the default value for this term in the 'experimental-theme.json' file.  The current schema used to determine this value is of the format `templateParts[ $slug ][ 'area' ]`.  For example, to define the 'area' of a template part whose file is named "file-name.html" the json would look like:
```
"templateParts": {
		"file-name": {
			"area": "header"
		}
	}
```
* Adds this term to the response for file based and custom template parts.
* Updates the rest controller to save this term in standard CRUD operations.
* Enables querying by this term.
* Restricts this term to a list of supported types.

## How has this been tested?

Many of the following steps are covered in the added php unit tests.  To test manually:

* Run this PR
### 1 - Verify the term is added on singular requests and persists across saving.
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
* Add this json code to the experimental-themes.json file of a block based theme:
```
"templateParts": {
		"header": {
			"area": "header"
		}
	}
```
* Load the editor with a 'fresh' theme install (no customized theme templates or template parts).
* Log the response for the single template part request (such as `entityRecord` [here](https://github.com/WordPress/gutenberg/blob/a0c91bcdfd8f86846a33227ba9851d84476e6bff/packages/block-library/src/template-part/edit/index.js#L49)) and verify the header's "area" key is present and has the expected value "header".
* Similarly, verify the footer (which you did not add any json for) is listed as "uncategorized".
* Make changes to the template part and save.
* Log the response again for the now custom template part (is_custom should now be `true`), and verify the term key/value is still present as expected.  

### 2 - Verify querying by the term works for both non-custom and custom template parts.
* Clean/delete the customized theme template parts from the testing step above.
* The theme's JSON file will still need the addition mentioned above.
* Request template parts querying `area: "header"` and log the result.  One way to do this is to add a query to [this request](https://github.com/WordPress/gutenberg/blob/69027a01c609627e68e663e521ee397dc584c2d3/packages/block-library/src/template-part/edit/selection/template-part-previews.js#L209-L213) and log the result.  This request can be triggered by pressing the "down" chevron in a template part block's toolbar (to the right of the renaming input).
```
	select( 'core' ).getEntityRecords( 'postType', 'wp_template_part', {
		area: 'header',
	} ) 
```
* Load the editor, trigger and log the request and verify only the header template part is returned in result.
* Edit the header template part and save.
* Again, trigger and log the response for the now custom template part  (is_custom should now be `true`) and verify it is the only result.

### 3 - Verify this does not work for unsupported types.
* In the theme's JSON file, change the area from type `header` to type `foobar`.
* Clean/delete custom theme template parts.
* Verify this unsupported area type is not added to the template part and shows up as `uncategorized`.

### 4 - Smoke test the normal template part flows.
* Undo any changes to Gutenberg made in the testing steps above (queries/logs).
* Verify the nav sidebar and template part selection show both header and footer as expected.
* Verify template part flows work as expected.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

Addison-Stavlo committed:
register term

Addison-Stavlo committed:
update rest controller and build result functions

Addison-Stavlo committed:
dont add term to regular template posts

Addison-Stavlo committed:
only prepare term for database if template part

youknowriad commented on lib/full-site-editing/block-templates.php:
```diff
@@ -173,6 +173,17 @@ function _gutenberg_build_template_result_from_file( $template_file, $template_t
+			true
+		);
+		if ( isset( $data['template-parts'][ $template_file['slug'] ]['type'] ) ) {
+			$template->template_part_type = $data['template-parts'][ $template_file['slug'] ]['type'];
```
It seems type might be redundunt with the "type" above. I wonder if we could find a better for this property. "scope" maybe? cc @mtias @mcsf

mtias replied:
I'm leaning towards "section"

Addison-Stavlo replied:
After stepping away I starting thinking `semantic_type` would fit better for both "type" and "template_part_type".  But if we end up allowing types such as "left sidebar" or things that don't really line up with semantic HTML elements 1-1, something like "section" seems fitting.

youknowriad reviewed (commented):
This PR is a great start, it matches very closely what I had in mind for this feature.

Addison-Stavlo committed:
rename 'section'

Addison-Stavlo committed:
add query support

Addison-Stavlo committed:
refactor json attribute injection, use for both plural and singular file getters

Addison-Stavlo committed:
only allow for supported section types

Addison-Stavlo committed:
remove unnecessary newlines

Addison-Stavlo commented on lib/full-site-editing/template-parts.php:
```diff
@@ -60,6 +60,35 @@ function gutenberg_register_template_part_post_type() {
+		'section',
+		array( 'wp_template_part' ),
+		array(
+			'public'            => false,
+			'hierarchical'      => false,
+			'labels'            => array(
+				'name'          => __( 'Sections', 'gutenberg' ),
+				'singular_name' => __( 'Section', 'gutenberg' ),
+			),
+			'query_var'         => false,
+			'rewrite'           => false,
+			'show_ui'           => false,
+			'_builtin'          => true,
+			'show_in_nav_menus' => false,
+			'show_in_rest'      => false,
```
Im not 100% sure on which settings we really want currently.  For now this is the same for how the `wp_theme` term is set up.  Maybe we would want this `public` and `show_ui` in the future? 🤔

jeyip replied:
I think these are good callouts. I'm not certain either.

Unless someone has strong use cases, I'd defer to hiding the ui, especially since there's little to no documentation about the subject. If I'm not mistaken, it's also something we could easily introduce in the future without causing breaking changes.

Addison-Stavlo replied:
>  I'd defer to hiding the ui, especially since there's little to no documentation about the subject. If I'm not mistaken, it's also something we could easily introduce in the future without causing breaking changes.

Thats what I was thinking as well.  Start with these settings (as long as they make sense to everyone) and update them in the future if necessary.

vindl replied:
One argument in favor of `show_ui` is that it might make development and debugging easier, since we can more easily inspect the values or manually add new ones there etc. Other than that I think this should be hidden. Also I was expecting that we'd need `show_in_rest` to be `true` down the road?

mtias replied:
We want this to be essentially hidden terms we select through the editor UI alone (similar to how post formats are a taxonomy) so we should disable most things.

Addison-Stavlo committed:
help function for ensuring supported section type

jeyip commented:
## Testing
### Behavior
- [x] Verify the term is added on singular requests and persists across saving.
- [x] Verify querying by the term works for both non-custom and custom template parts.
- [x] Verify this does not work for unsupported types.
- [x] Smoke test the normal template part flows.

### Browsers
Will continue testing in other browsers tomorrow.

- [x] Chrome
- [ ] Firefox
- [ ] IE11
- [ ] Edge
- [ ] Safari

Addison-Stavlo committed:
fix php lint error

jeyip commented on lib/full-site-editing/block-templates.php:
```diff
@@ -93,18 +95,42 @@ function _gutenberg_get_template_files( $template_type ) {
+ *
+ * @return array Template.
+ */
+function _gutenberg_add_template_part_section_info( $template_info ) {
```
`Non-blocking:`

The name of this method confused me a little. When I first saw it on l52 and l98, I thought that we were _always_ appending template part section info. In reality, we add it conditionally.  

Curious how we would feel about breaking down `_gutenberg_add_template_part_section_info` into two methods with distinct responsibilities and refactoring like so:

```php
function _gutenberg_get_template_file( $template_type, $slug ) {
                         // only included relevant code in this method
			$template_info = array(
				'slug'  => $template_slug,
				'path'  => $template_file,
				'theme' => $theme_slug,
				'type'  => $template_type,
			)

			if ( _should_add_section_info( $template_type )) {
				$template_files[] = _gutenberg_add_template_part_section_info($template_info);
			} else {
				$template_files[] = $template_info;
			}
		}
	}

	return $template_files;
}

function _should_add_section_info( $template_type ) {
	return  'wp_template_part' === $template_type && is_readable( locate_template( 'experimental-theme.json' ) )
};

function _gutenberg_add_template_part_section_info( $template_info ) {
	$theme_json = file_get_contents( locate_template( 'experimental-theme.json' ) );
	$data       = json_decode(
		$theme_json,
		true
	);
	if ( isset( $data['template-parts'][ $template_info['slug'] ]['section'] ) ) {
		$template_info['section'] = gutenberg_filter_template_part_section_type( $data['template-parts'][ $template_info['slug'] ]['section'] );
	}

	return $template_info;
}
```

Addison-Stavlo replied:
I was scratching my head over the best way to go about this as well.  That may make sense!  Initially, I wanted to avoid having to put the same conditional logic for whether or not to add the section info in both `_gutenberg_get_template_file` and `gutenberg_get_template_files` which is why I moved it all into that add function. 


Refactoring it like this means we will have that same if/else in both single/plural versions of the get functions, but thats not necessarily the end of the world... I just noticed myself typing the same thing in 2 places and wanted to consolidate it.

Do you think it makes more sense to separate this into 2 functions, or to rename the function to `*_conditionally_add_*`? 🤷‍♀️

jeyip replied:
> Refactoring it like this means we will have that same if/else in both single/plural versions of the get functions, but thats not necessarily the end of the world... I just noticed myself typing the same thing in 2 places and wanted to consolidate it.

Ah that makes sense.

> Do you think it makes more sense to separate this into 2 functions, or to rename the function to *_conditionally_add_*? 🤷‍♀️

Hmm...I'm thinking about other names as well. I do think it would be more informative than what we currently have.

> I wanted to avoid having to put the same conditional logic for whether or not to add the section info in both _gutenberg_get_template_file and gutenberg_get_template_files which is why I moved it all into that add function.
One other option is to extract the conditional logic in both 

One other option is extracting the similar conditional logic into its own method as well 🤔

```php
function _gutenberg_get_template_file( $template_type, $slug ) {
                         // only included relevant code in this method
			$template_info = array(
				'slug'  => $template_slug,
				'path'  => $template_file,
				'theme' => $theme_slug,
				'type'  => $template_type,
			)

                        $template_files[] = _generate_template_file_from_info($template_info);
		}
	}

	return $template_files;
}

function _gutenberg_get_template_files( $template_type ) {
                         // only included relevant code in this method
			$template_info = array(
				'slug'  => $template_slug,
				'path'  => $template_file,
				'theme' => $theme_slug,
				'type'  => $template_type,
			)

                        $template_files[] = _generate_template_file_from_info($template_info);
		}
	}

	return $template_files;
}
```

I think renaming it _or_ creating an additional method to DRY out conditional logic would be good

Addison-Stavlo replied:
Im not totally following the last example.  It looks like what is happening now with just the function renamed (same input, used in same place.)?

jeyip replied:
I updated the last example to include more information. I should've clarified a few things:

> I wanted to avoid having to put the same conditional logic for whether or not to add the section info in both _gutenberg_get_template_file and gutenberg_get_template_files which is why I moved it all into that add function.

The last example builds off of the first example. In the first example, we'd end up with duplicated conditional logic in `_gutenberg_get_template_file` and `gutenberg_get_template_files`. In the last example, we're adding one additional method called `_generate_template_file_from_info` to extract the duplicated conditional logic into its own method. Does that make more sense?

jeyip replied:
Oh I see what you're saying! Yes the solution ends up matching the structure of what you originally had, except with a few more declarative helper methods.

Addison-Stavlo replied:
Gotcha, thanks for clarifying.  I see what you are saying.  This would make sense to me if these helper methods had a use elsewhere, but Im not sure its worth splitting up into 3 functions for short helper methods that will only be used in 1 place.  With that in mind id be more inclined to just rename the original function with `*_conditionally_*`.

jeyip replied:
b1138e0 lgtm!

jeyip reviewed (commented):
Everything looks reasonable 👍 . The new template part section taxonomy appears to behave as expected.

+1 from me whenever we add php unit specs to the PR

Addison-Stavlo committed:
rename function

david-szabo97 commented on lib/full-site-editing/block-templates.php:
```diff
@@ -93,18 +95,42 @@ function _gutenberg_get_template_files( $template_type ) {
+ * @return array Template.
+ */
+function _gutenberg_conditionally_add_template_part_section_info( $template_info ) {
+	if ( 'wp_template_part' === $template_info['type'] && is_readable( locate_template( 'experimental-theme.json' ) ) ) {
```
I'm surprised we don't have any functions to check for the theme.json file... and we don't have any functions to read it 🤔 It would be great to load the file only once per request. We are checking and reading, decoding the file for each template part.

We have a function called `gutenberg_experimental_global_styles_has_theme_json_support`, which checks for the existence for the `experimental-theme.json` file. Later we should remove the `global_styles` prefix and move it to a different file so it acts as a global function for checking for theme JSON support. (It's not something we should tackle in this PR, just a thought)

mtias replied:
Not sure why we would be checking for theme.json here?

Addison-Stavlo replied:
> Not sure why we would be checking for theme.json here?

@mtias - In order to allow themes to define a starting value for the term for the template parts they provide.  @youknowriad suggested using the theme.json and it seems like a good way to go about it.  Alternatives i could think of  would be going by slug (filename) prefix ( `header-*` ) or maybe adding some special markup in the file that doesnt correspond to an element or block.. but it seems like allowing them to define the term in the theme.json seems best of those options.  Open to other ideas though!

> It would be great to load the file only once per request. We are checking and reading, decoding the file for each template part.

@david-szabo97 good point!  I think that would make sense for now, but if our goal in the future is to start supporting items from non-active themes that might be a bit more difficult.  Will see if I can update this to only check and decode it once.

Addison-Stavlo replied:
> We are checking and reading, decoding the file for each template part.

Only decoding once per request now. 👍

david-szabo97 reviewed (commented):
Looking good and works as expected! Great work there! Looking forward to the tests 😄

Copons commented:
I wonder if "section" is the right term for this. 🤔 

Somehow my brain is stuck thinking that "location" might be more appropriate, as a kind of callout to [`register_nav_menu`](https://developer.wordpress.org/reference/functions/register_nav_menu/).

Neither sound perfect, though. Maybe we already have a WordPress-y term that fits this use-case?

Addison-Stavlo commented:
> I wonder if "section" is the right term for this. 🤔

cc @Copons  -  Yeah, I'm not certain what the best name is either.  We were discussing this in the above comments (https://github.com/WordPress/gutenberg/pull/28410#discussion_r562429813), which I 'unresolved' for now if we want to keep discussing better alternatives.

mattwiebe commented:
The closest prior art in WP for this is Widget Areas, itself an attempt at better naming for Sidebars that were no longer always on the side, or a bar, for that matter.

So I'd suggest `wp_template_area` for the taxonomy, or whatever variation of `wp_` and/or `template_` prefixes we think are desirable (I would use both to maintain the ties to `wp_template` and `wp_template_part`).

The main reason I do like `section` (but with `wp_template_section` as the full name) is that it maps to the HTML semantic `<section>` tag. 

Otherwise **zone** and **region** are nice standouts from my trip to the thesaurus.

Addison-Stavlo committed:
only decode json once per request

Addison-Stavlo committed:
update 'from post' test

Addison-Stavlo committed:
rename term 'wp_template_section'

Addison-Stavlo committed:
update rest tests for template parts

Addison-Stavlo committed:
update block templates tests

Addison-Stavlo committed:
add TODOs to follow up on after updating themes

Addison-Stavlo commented:
@mattwiebe -  That makes sense to me!  Updated now so we register the term as `wp_template_section` but we can still query and handle json with just `section` (sort of how templates and template parts have the `theme` attribute for the `wp_theme` term).  Do you think that makes sense?

Addison-Stavlo commented:
Updated per feedback:
* changed the registered tax term.
* made sure we are not decoding json more than once per request.
* added template part specific tests using this term in block-templates and rest controller tests.

david-szabo97 commented on lib/full-site-editing/block-templates.php:
```diff
@@ -46,15 +46,30 @@ function _gutenberg_get_template_file( $template_type, $slug ) {
+					$new_template_item,
+					$template_part_data
+				);
+			} else {
```
I'm not sure what's the convention in the PHP part of Gutenberg, but I usually drop the `else` branch in this case and just leave the return on a newline. Since we are returning in the `if` branch we are safe to drop the `else` branch.

Addison-Stavlo replied:
Lol very true, im not sure what I was thinking when I put the `else` in.

jeyip replied:
Yeah Marko pointed this out in [one of my PRs](https://github.com/WordPress/gutenberg/pull/28088#discussion_r557733884) as well. I think it's an unspoken convention to remove the return from an `else` branch (maybe worth making this an actual linting rule)

Addison-Stavlo replied:
I think not having the `else` makes the most sense and reads well.  I could swear I've seen a lint rule for that somewhere (maybe in JS or in another repo).  Anyways, updated!

david-szabo97 commented on lib/full-site-editing/block-templates.php:
```diff
@@ -269,8 +352,9 @@ function gutenberg_get_block_templates( $query = array(), $template_type = 'wp_t
 			);
 			$should_include = false === $is_custom && (
-				! isset( $query['slug__in'] ) || in_array( $template_file['slug'], $query['slug__in'], true )
-			);
+				! isset( $query['slug__in'] ) || in_array( $template_file['slug'], $query['slug__in'], true ) ) && (
+					! isset( $query['section'] ) || $template_file['section'] === $query['section']
+				);
```
Duh, this is difficult to read. Maybe we should split this into a few variables?

Addison-Stavlo replied:
Updated, should be more clear now.

david-szabo97 commented on phpunit/class-block-templates-test.php:
```diff
@@ -55,6 +78,25 @@ function test_gutenberg_build_template_result_from_file() {
+		$this->assertEquals( 'header', $template_part->title );
+		$this->assertEquals( '', $template_part->description );
+		$this->assertEquals( 'wp_template_part', $template_part->type );
+		// TODO - update null to 'header' once tt1-blocks theme.json updated for template part section info.
```
Feels weird that we are depending on the TT1B theme for unit tests. Isn't there a way to create a theme fixture and use that for unit testing? (Not a blocker, but it would be a great follow up)

Addison-Stavlo replied:
> Isn't there a way to create a theme fixture and use that for unit testing?

🤔  Im not sure the best way to go about that.  It seems like it would need to be installed and activate-able like a regular theme, with the same file structures.  And at that point maybe using TT1B just makes the most sense (as opposed to creating and maintaining a separate one just for testing) since it is the most kept up-to-date block-based theme that we are developing FSE around currently.  I wonder if @youknowriad has given this any thought previously?

david-szabo97 replied:
It makes more sense for E2E tests. I don't think it's reasonable for unit tests. We opened an issue to discuss it: https://github.com/WordPress/gutenberg/issues/28692 Would love to hear your thoughts. It's not a blocker, just a thought 😄

david-szabo97 reviewed (approved):
Looks good to me ✅ Left a few non-blocker (nitpicking) comments/feedbacks 😄

mattwiebe commented:
> Updated now so we register the term as `wp_template_section` but we can still query and handle json with just `section` (sort of how templates and template parts have the `theme` attribute for the `wp_theme` term).  Do you think that makes sense?

Yup, that sounds perfect.

Addison-Stavlo committed:
rename register function

Addison-Stavlo committed:
remove unnecessary else

Addison-Stavlo committed:
fix renamed function in tests

Addison-Stavlo committed:
make should_include logic more readable

vindl commented on lib/full-site-editing/block-templates.php:
```diff
@@ -93,18 +112,62 @@ function _gutenberg_get_template_files( $template_type ) {
+	if ( isset( $data['template-parts'] ) ) {
+		return $data['template-parts'];
+	}
+	return array();
```
Should we add newline before return here to make it consistent with the rest of the formatting? (same goes for if block above and a couple of other changes in this PR)

vindl commented on lib/full-site-editing/block-templates.php:
```diff
@@ -239,6 +312,15 @@ function gutenberg_get_block_templates( $query = array(), $template_type = 'wp_t
 	);
 
+	if ( 'wp_template_part' === $template_type && isset( $query['section'] ) ) {
+		$wp_query_args['tax_query'][]           = array(
```
Aligning `=` looks off here. :)

Addison-Stavlo replied:
Composer formatted that way to line up with the `relation` defined just below.

vindl replied:
I figured, some of its auto-formatting is just weird. :D

vindl commented on lib/full-site-editing/template-parts.php:
```diff
@@ -118,3 +147,38 @@ function set_unique_slug_on_create_template_part( $post_id ) {
+ */
+function gutenberg_get_allowed_template_part_section_types() {
+	$default_section_types = array(
+		'header',
```
What do you think about introducing these as constants? It might be beneficial down the road for comparisons, autocomplete, avoiding bugs due to typos etc.

Addison-Stavlo replied:
Im not sure I follow exactly what you are suggesting, but it sounds like it would make sense.

vindl replied:
I'm thinking something along the lines of:

define( `TEMPLATE_PART_SECTION_HEADER', 'header' );

and then using that throughout the code instead. It probably needs to be prefixed with `gutenberg` too, at which points it becomes somewhat unwieldy. Which makes me think that we could isolate this code within `WP_Template_Part` class to avoid the need for prefixing functions and constants all the time.

Addison-Stavlo replied:
defined them as constants here - d39b70f.  Im not sure if its necessary to check `defined(` first, but it seems like it couldn't hurt.

vindl reviewed (commented):
Great start here! I like the approach with storing the section info in `theme.json`.

vindl commented on lib/full-site-editing/template-parts.php:
```diff
@@ -118,3 +147,38 @@ function set_unique_slug_on_create_template_part( $post_id ) {
+	if ( in_array( $type, gutenberg_get_allowed_template_part_section_types(), true ) ) {
+		return $type;
+	}
+	return 'other';
```
What are the benefits of defaulting un-allowed types to `other` instead of leaving them undefined and maybe showing an error message or a warning?

Addison-Stavlo replied:
Maybe its not necessary, but it allows a distinction between those that are uncategorized vs. those that are intended for a specific purpose but categorized by an unsupported term -  template parts who had no section defined at all `null`/`undefined` vs. template parts who are intended for a particular section that happens to be unsupported `'other'`.

I suppose we should probably trigger a warning here as well?

Addison-Stavlo replied:
Set up a warning for this - 8c5f17b

Addison-Stavlo commented on lib/full-site-editing/block-templates.php:
```diff
@@ -93,18 +113,62 @@ function _gutenberg_get_template_files( $template_type ) {
+ * @return array Template part data from theme.json or empty array if not found.
+ */
+function _gutenberg_get_template_part_info_from_theme_json() {
+	if ( is_readable( locate_template( 'experimental-theme.json' ) ) ) {
```
I think the advantage of `locate_template` is that is accounts for parent themes and falls back to TEMPLATEPATH after STYLESHEETPATH.  This is similar to how global styles checks the theme JSON:

https://github.com/WordPress/gutenberg/blob/537412fd439c0c8a9befcf99e8bbbc2fdc8c1147/lib/global-styles.php#L13-L15

Now looking back, it seems like we should probably be using that global styles function here?

I don't think our problem with STYLESHEETPATH as noted in that issue exists here, since this gets called much later in the process than that example. 🤔

vindl replied:
I think the behavior should be the same in both cases. The benefits of not using `locate_template` would be:

- not calling the old code that's supposed to handle legacy themes (not applicable to FSE themes);
- `get_stylesheet_directory` is filterable. 

> Now looking back, it seems like we should probably be using that global styles function here?

Agreed. I think it would be better to reuse the existing global styles function. I'm still thinking that it could be refactored to use `get_stylesheet_directory` too (cc @nosolosw)

nosolosw replied:
:wave: I'd like to share a different perspective on data should be read out of theme.json. This is how I envision it working:

- Every new field is added to [the schema](https://github.com/WordPress/gutenberg/blob/master/lib/class-wp-theme-json.php#L105) we have.
- The class `WP_Theme_JSON` is used to get data out of it. If necessary, by creating new getter methods. Example: `get_section` (just an example, not sure if it's a good name yet).

This approach has many benefits, including that we only read the theme.json once (I/O operation, costly) and that the interaction with the underlying data is abstracted away (keys can be moved to a different place without having to touch multiple files, the hard parts like merging theme/user data is done automatically for all fields that need it, etc).

As a follow-up to the recent refactors https://github.com/WordPress/gutenberg/pull/28110 https://github.com/WordPress/gutenberg/pull/28533 I had in my queue create a PR to migrate the existing `customTemplates` to work like that. Happy to accelerate that PR if it serves as inspiration for this one.

Addison-Stavlo replied:
>The class WP_Theme_JSON is used to get data out of theme.json. If necessary, by creating new getter methods. Example: get_section (just an example, not sure if it's a good name yet).

I think that makes a lot of sense!  Will see if I can get something like that working.  So we have an instance of `WP_Theme_JSON` already set up and just use a get method on it to get our information (instead of us having to decode in the request here)?

>Happy to accelerate that PR if it serves as inspiration for this one.

Im always open for more inspiration! 😆  - But no pressure if you have other things going on.

nosolosw replied:
ok, took a quick stab at this and prepared https://github.com/WordPress/gutenberg/pull/28700 

I still need to understand how `pageTemplates` is being used and whether other things need to be added, but it looks promising.

Addison-Stavlo replied:
@nosolosw - Thanks!  So something along the lines of this: 0659325 ?  (although eventually using the `get_theme_data` you are proposing in that PR as opposed to just using `get_theme_origin` and making it public like im doing here).

nosolosw replied:
Yeah. Would you help me review / land https://github.com/WordPress/gutenberg/pull/28700 so you can have it for this as well? I can polish it this morning.

nosolosw replied:
https://github.com/WordPress/gutenberg/pull/28700 is ready to land!

Addison-Stavlo replied:
So we did update this to work the JSON resolver which is great, but I think we might be getting a little distracted from @vindl 's original critique.  I think that was that the use of `locate_template` (and underlying `STYLESHEETPATH`) used in the json_resolver class, as opposed to using things like `get_stylesheet_path` that are filterable.  (Please free to clarify @vindl ).. But do you have thoughts regarding that @nosolosw ?

nosolosw replied:
ok, loaded my brain with these conversations. I see this comes from https://github.com/WordPress/gutenberg/pull/26500#discussion_r515047360 which ended up being problematic, hence why we did https://github.com/WordPress/gutenberg/pull/27021 What I've gathered is that the use of non-filterable constants (`STYLESHEET` instead of `get_stylesheet_directory`) is problematic under certain special setups, hence why `locate_template` is best to be avoided.

The use of `locate_template` in global styles boils down to support child themes that want to reuse the parent's `theme.json`. So, the change involves replicating part of the logic of `locate_template` but with methods instead of constants. Probably not necessary to have the theme-compat part. I can do that after https://github.com/WordPress/gutenberg/pull/28855 lands ― it needs a review 🙂

nosolosw replied:
Here's a bug report https://github.com/WordPress/gutenberg/issues/26359 that confirms that `locate_template` is problematic in certain contexts (customizer).

nosolosw replied:
https://github.com/WordPress/gutenberg/pull/28942 replaces `locate_template`.

nosolosw replied:
just need a :+1: to land https://github.com/WordPress/gutenberg/pull/28942 :)

Addison-Stavlo commented on lib/full-site-editing/block-templates.php:
```diff
@@ -93,18 +113,62 @@ function _gutenberg_get_template_files( $template_type ) {
+ *
+ * @return array Template.
+ */
+function _gutenberg_conditionally_add_template_part_section_info( $template_info, $theme_data ) {
```
>I've seen a couple of usages of leading underscore through Gutenberg, but I don't think it's documented in WP coding conventions. Do we really need it here?

I noticed some of the other functions in this file use the leading underscore as well.  From inspection I am assuming helper functions only intended to be called in this file have the leading underscore, while more robust functions also used elsewhere do not.

>Also I think we could drop conditionally from the name of the function since I don't think it helps much with readability.

👍  Yeah, after refactoring that no longer seems necessary.

Addison-Stavlo commented on lib/full-site-editing/block-templates.php:
```diff
@@ -208,6 +276,12 @@ function _gutenberg_build_template_result_from_post( $post ) {
 
+	if ( 'wp_template_part' === $post->post_type ) {
+		$type_terms        = get_the_terms( $post, 'wp_template_section' );
+		$section           = $type_terms[0]->name;
```
> Also get_the_terms can return false or WP_error and we are not handling those cases here?

Oh, good point!  I wonder if this also needs to be updated for this pre-existing bit as well:

https://github.com/WordPress/gutenberg/blob/537412fd439c0c8a9befcf99e8bbbc2fdc8c1147/lib/full-site-editing/block-templates.php#L345

Addison-Stavlo replied:
How do you think we should handle the error case here?  Returning the error instead of the template probably doesn't make sense since `section` is not required.. we probably just want to not add the section field and continue to return the template?  or?

vindl replied:
> we probably just want to not add the section field

I'm leaning towards this.

Addison-Stavlo replied:
Cool, I updated ours to work that way.  Im still a bit unsure about what to do about that pre-existing `get_the_terms` I noted above that is used in `gutenberg_filter_wp_template_unique_post_slug`.

Addison-Stavlo commented on lib/full-site-editing/block-templates.php:
```diff
@@ -93,18 +113,62 @@ function _gutenberg_get_template_files( $template_type ) {
+ */
+function _gutenberg_get_template_part_info_from_theme_json() {
+	if ( is_readable( locate_template( 'experimental-theme.json' ) ) ) {
+		$theme_json = file_get_contents( locate_template( 'experimental-theme.json' ) );
```
also same as above 😁

Addison-Stavlo committed:
update format, function name, and check terms error

mcsf commented on phpunit/class-wp-rest-template-controller-test.php:
```diff
@@ -103,6 +129,32 @@ public function test_get_item() {
 		);
+
+		// Test template parts.
+		$request  = new WP_REST_Request( 'GET', '/wp/v2/template-parts/tt1-blocks//header' );
```
Not a blocker, but could you point me to where `theme-name//part-slug` was decided as an identifier?

Addison-Stavlo replied:
Sure thing!  I believe that was introduced in Riad's previous PR to remove auto-drafts https://github.com/WordPress/gutenberg/pull/27910 .

Im having trouble finding the exact conversation there, but I think it was originally `theme|slug` and @Copons had made an argument to update it to `theme//slug` IIRC.

Addison-Stavlo replied:
https://github.com/WordPress/gutenberg/pull/27910#discussion_r553288550

mcsf reviewed (commented):
I'm still trying to decide where I stand on using theme.json for provisioning template parts' terms, but the work here is coherent.

Addison-Stavlo commented:
@mcsf - thanks for taking a look!

> I'm still trying to decide where I stand on using theme.json for provisioning template parts' terms, 

It does seem a bit heavy handed, doesn't it?  Do you have any alternative ideas?

At this point it seems better than the alternatives I could think of.  Other thoughts were inferring it from the slug/file-name.  That would require everyone to name template part files like `header-*` or some similar consistent format.  Another option might be to check for some special markup in the files themselves that doesn't correspond to blocks or elements and serve the purpose of conveying this information for the tax term.  Interested in other ideas or suggestions!

nosolosw commented:
:wave: I see this PR is already approved but wanted to surface my perspective on how we should interact with data stored in `theme.json`: https://github.com/WordPress/gutenberg/pull/28410/files#r569561199

Would like to hear your thoughts on that before landing this.

Addison-Stavlo committed:
define constants

Addison-Stavlo committed:
trigger warning when unsupported type is used

Addison-Stavlo committed:
use WP_Theme_JSON_Resolver

Addison-Stavlo committed:
just some spacing

mtias commented on lib/full-site-editing/template-parts.php:
```diff
@@ -60,6 +60,49 @@ function gutenberg_register_template_part_post_type() {
+	}
+
+	register_taxonomy(
+		'wp_template_section',
```
Would it make more sense to use `wp_theme_section` here to align with the other relevant taxonomy we register (`wp_theme`)?

Addison-Stavlo replied:
Im not sure.  `wp_theme_section` seems odd as its not the section of a theme that is being defined, just the area in a template that the template part is designated for.  `wp_theme` is theme relevant because it links the cpt to the theme and is the theme name/directory, while `wp_theme_section` seems odd because the theme is entirely irrelevant with respect to the section (other than just defining the section for the template parts they supply).

Addison-Stavlo replied:
^^ That said, im open to what others think makes the most sense.

mtias replied:
Yes, I'm not sure myself :) Maybe we just need to drop the "template" for `wp_section`, but then there's some higher likelihood of clashes with possible third-party definitions?

mtias commented on lib/full-site-editing/template-parts.php:
```diff
@@ -60,6 +60,49 @@ function gutenberg_register_template_part_post_type() {
+			'hierarchical'      => false,
+			'labels'            => array(
+				'name'          => __( 'Sections', 'gutenberg' ),
+				'singular_name' => __( 'Section', 'gutenberg' ),
```
Might be worth appending "Theme Section" here for clarity

mtias commented on lib/full-site-editing/template-parts.php:
```diff
@@ -60,6 +60,49 @@ function gutenberg_register_template_part_post_type() {
+	define( 'WP_TEMPLATE_SECTION_SIDEBAR', 'sidebar' );
+}
+if ( ! defined( 'WP_TEMPLATE_SECTION_OTHER' ) ) {
+	define( 'WP_TEMPLATE_SECTION_OTHER', 'other' );
```
Not super sure "other" should be defined. Should we instead go with just plain uncategorized?

Addison-Stavlo replied:
Maybe that makes more sense.  My initial thought was that TPs that dont have a section term defined yet (such as the theme not supplying the term for their TPs, or the cpt being created elsewhere where the section is not specified) would not have a section set and thus correspond to being uncategorized.  While TPs that are created with a section term that is unsupported (like 'diagonal sidepanel' as a goofy example) would be categorized as 'other'.  So 'other' being different than uncategorized as: "we tried to categorize it, but it didnt make sense."

Do you think it makes more sense for both cases to just fall under uncategorized together?  Maybe thats a more simple way to go about it... im definitely open to doing it that way instead.

Addison-Stavlo replied:
Updated so both cases just go with uncategorized.

mtias replied:
Thanks, that seems like a better start. It's always easier to add more terms / concepts later than to remove them.

Addison-Stavlo committed:
use new methods after rebase updates

Addison-Stavlo committed:
use 'uncategorized' for both unspecified and unsupported

Addison-Stavlo committed:
remove unnecessarily added newlines

Addison-Stavlo commented:
Updated per feedback:
* Incorporated recent changes to use the WP_Theme_JSON_Resolver (thanks @nosolosw!)
* Got rid of the "other" category and updated so unspecified and unsupported sections are added as "uncategorized".

Addison-Stavlo committed:
add test for new theme json getter

lukecarbis commented:
@mattwiebe I personally don't like `section` _because_ it is known HTML markup. Especially in context with `header` as given above, it could get quite confusing. Is this `header` a template `section` or `<section>` or `<header>`?

I like **zone**, **region**, and **location**.

mtias commented:
Naming is always the hardest part :)

nosolosw commented on lib/class-wp-theme-json.php:
```diff
@@ -1064,6 +1065,18 @@ public function get_page_templates() {
+	 *
+	 * @return array
+	 */
+	public function get_template_part_data() {
```
What if we rename this to `get_template_parts` so the getter mimics the field name?

```suggestion
	public function get_template_parts() {
```

nosolosw commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -759,4 +759,27 @@ function test_get_page_templates() {
 		);
 	}
+
+	function test_get_template_part_data() {
```
```suggestion
	function test_get_template_parts {
```

Addison-Stavlo committed:
Update lib/class-wp-theme-json.php

Co-authored-by: André <nosolosw@users.noreply.github.com>

Addison-Stavlo committed:
Update phpunit/class-wp-theme-json-test.php

Co-authored-by: André <nosolosw@users.noreply.github.com>

nosolosw commented:
@Addison-Stavlo in https://github.com/WordPress/gutenberg/pull/28830/commits/4019f145acf6fd8b6daf2495008a6d029ef5e367 I've started to add docs for the "metadata" keys that we're adding to theme.json, so it's documented. If https://github.com/WordPress/gutenberg/pull/28830 lands before this, this could be perhaps rebased to add the new `templateParts` field in the docs. Can be done in a follow-up PR as well.

Another thing that can be also a follow-up is to think about how this field needs validation (or not). I understand `section` is mandatory for the `templateParts` declared. What if we:

- added validation to not save the parts that don't have that field? They won't be listed in `get_template_parts`.
- other idea is: perhaps we can pre-fill that field with the `WP_TEMPLATE_SECTION_UNCATEGORIZE` value if it doesn't have one? This way we don't need to replicate that page logic anywhere else and can be expressed in tests.

Addison-Stavlo committed:
update method name calls

Addison-Stavlo commented:
> Is this header a template section or <section> or <header>?

@lukecarbis -  I see what you are saying, its a bit unfortunate naming `"section": "header"` - is the 'section' a <header> or is the 'header' a <section>?

>Naming is always the hardest part :)

@mtias  Whats the best way to come to a consensus on what makes the most sense? 😆   area, zone, region, semanticType, anything else?  They all seem like they would work.

> I understand section is mandatory for the templateParts declared.

I guess that depends on what we mean by 'mandatory'.  Currently its the only field we use so it doesn't make much sense without it, but maybe we should support the theme providing excerpt, description, or other fields here in the future.  However, the not required for the theme to add the section field as we would always just fall back to uncategorized.  Right now we are applying the uncategorized on the block templates end, but as you not maybe that makes more sense to move that as you noted 🤔 :

> other idea is: perhaps we can pre-fill that field with the WP_TEMPLATE_SECTION_UNCATEGORIZE value if it doesn't have one?

mtias commented:
Let's go with "areas":

![image](https://user-images.githubusercontent.com/548849/107372196-6e914000-6ae5-11eb-8db7-5e2c2d1ab046.png)

Addison-Stavlo committed:
rename section to area

Addison-Stavlo commented:
> Let's go with "areas":

👍  Sounds good. Updated! ✅

david-szabo97 commented:
Tested according to the testing instructions and smoke tested the editor. Everything works fine! ✅

jeyip commented:
## Testing
### Behavior
Followed the instructions and everything is working as expected with the new area changes.

- [x] Verify the term is added on singular requests and persists across saving.
- [x] Verify querying by the term works for both non-custom and custom template parts.
- [x] Verify this does not work for unsupported types.
- [x] Smoke test the normal template part flows.

jeyip commented:
Oh! My mistake. I am actually seeing something potentially strange.

I added a new template part to the tt1-blocks index template in the site editor. The area value seems strange `"WP_TEMPLAT_PART_AREA_UNCATEGORIZED"`

<img width="440" alt="Screen Shot 2021-02-10 at 6 47 30 AM" src="https://user-images.githubusercontent.com/5414230/107525627-e30ec200-6b6b-11eb-9540-8ca148eb2cff.png">

Copons commented:
> Oh! My mistake. I am actually seeing something potentially strange. +1 whenever we resolve this conversation.
> 
> I added a new template part to the tt1-blocks index template in the site editor. The area value seems strange `"WP_TEMPLAT_PART_AREA_UNCATEGORIZED"`

The typo is here:

https://github.com/WordPress/gutenberg/pull/28410/files#diff-805d551e09a2516b29027c017fcf01c1cf36e7131af223322df72ac22e755f7eR370

Wonder why PHP evaluates an undefined constant as a string literal, though... 😮

Addison-Stavlo committed:
fix typo

Addison-Stavlo commented:
> Oh! My mistake. I am seeing something potentially odd.
> The typo is here:

Oh good catch!  I assumed there would be typos from going through and renaming everything but thought I had caught them all.  Should be fixed now.

> Wonder why PHP evaluates an undefined constant as a string literal, though... 😮

TIL 🤷‍♀️

youknowriad commented on lib/full-site-editing/block-templates.php:
```diff
@@ -93,18 +98,45 @@ function _gutenberg_get_template_files( $template_type ) {
+	}
+
+	if ( isset( $theme_data[ $template_info['slug'] ]['area'] ) ) {
+		$template_info['area'] = gutenberg_filter_template_part_area_type( $theme_data[ $template_info['slug'] ]['area'] );
```
There's some inconsistency in different parts of the code base for instance in this function name`gutenberg_filter_template_part_area_type` we refer to the "area" as "type" which can be confusing especially since we already have a "type" property on the template and template area objects.

Addison-Stavlo replied:
Id say here we refer to 'area' as 'area', and 'type' more refers to the different allowed values for that area term.  Perhaps `gutenberg_filter_template_part_area_value` would make more sense?

youknowriad replied:
I like value better, but why not just `gutenberg_filter_template_part_area` ?

Addison-Stavlo replied:
that works too, names are hard 😅

Addison-Stavlo replied:
https://github.com/WordPress/gutenberg/pull/29679

gziolo commented:
Can you point me to the documentation that explains how to use `templateParts` in `theme.json`?

Can `area` hold a value that is translatable? Based on the example in the description of this PR, it doesn't need to be translated, but in the unit test I see "Some area" so I'm no longer sure about it.

Addison-Stavlo commented:
Hi @gziolo!

> Can you point me to the documentation that explains how to use templateParts in theme.json?

The example in the PR description should work, but here is how it is set up in tt1-blocks (https://github.com/WordPress/theme-experiments/blob/1a4128f7c2c741e9c14b4c2a6bffd9c5b06e77e2/tt1-blocks/experimental-theme.json#L2-L9).

> Can area hold a value that is translatable? 

I don't think translation makes sense here, as only specific values are allowed and these are all "under the hood".  When we surface an allowed value in the editor as an option, that is translated.

> Based on the example in the description of this PR, it doesn't need to be translated, but in the unit test I see "Some area" so I'm no longer sure about it.

I believe the thought behind that unit test is to ensure that if the value is registered as something that is not supported, that it gets turned into the 'uncategorized' value.  So "some area" is just an arbitrary unsupported string that should come get mapped to 'uncategorized'.

Addison-Stavlo commented:
editing comments seems a bit broken at the moment:

> Can you point me to the documentation that explains how to use templateParts in theme.json?

I think we need to add some proper documentation for this.  Do you know where the best place for that would be?

nosolosw commented:
> I think we need to add some proper documentation for this. Do you know where the best place for that would be?

A good first step would be to add it to the current theme.json [spec document](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/) ([source](https://github.com/WordPress/gutenberg/blob/trunk/docs/how-to-guides/themes/theme-json.md)).

Addison-Stavlo commented:
> A good first step would be to add it to the current theme.json spec document (source).

👍  - Thanks! That sounds like a good place to start - https://github.com/WordPress/gutenberg/pull/30118

mcsf commented on lib/full-site-editing/block-templates.php:
```diff
@@ -237,6 +280,15 @@ function gutenberg_get_block_templates( $query = array(), $template_type = 'wp_t
 		),
 	);
 
+	if ( 'wp_template_part' === $template_type && isset( $query['area'] ) ) {
```
`'area'` is a new key under `$query` and isn't documented in the `gutenberg_get_block_templates` docblock. @Addison-Stavlo, could you follow up with that quick addition? 🙏

Addison-Stavlo replied:
Certainly!  My apologies for the oversight there earlier - https://github.com/WordPress/gutenberg/pull/32746

mcsf replied:
Not at all, thank you!
